profiling/src/internal/profile/interning_api/mod.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright 2025-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+mod generational_ids;
+pub use generational_ids::*;
+
+use crate::api::ManagedStringId;
+use crate::collections::identifiable::{Dedup, StringId};
+use crate::internal::{
+    Function, FunctionId, Label, LabelId, LabelSet, LabelSetId, Location, LocationId, Mapping,
+    MappingId, Profile, Sample, StackTrace, StackTraceId, Timestamp,
+};
+use std::sync::atomic::Ordering::SeqCst;
+
+impl Profile {
+    pub fn intern_function(
+        &mut self,
+        name: GenerationalId<StringId>,
+        system_name: GenerationalId<StringId>,
+        filename: GenerationalId<StringId>,
+        start_line: i64,
+    ) -> anyhow::Result<GenerationalId<FunctionId>> {
+        let function = Function {
+            name: name.get(self.generation)?,
+            system_name: system_name.get(self.generation)?,
+            filename: filename.get(self.generation)?,
+            start_line,
+        };
+        let id = self.functions.dedup(function);
+        Ok(GenerationalId::new(id, self.generation))
+    }
+
+    pub fn intern_label_num(
+        &mut self,
+        key: GenerationalId<StringId>,
+        val: i64,
+        unit: Option<GenerationalId<StringId>>,
+    ) -> anyhow::Result<GenerationalId<LabelId>> {
+        let key = key.get(self.generation)?;
+        let unit = unit.map(|u| u.get(self.generation)).transpose()?;
+        let id = self.labels.dedup(Label::num(key, val, unit));
+        Ok(GenerationalId::new(id, self.generation))
+    }
+
+    pub fn intern_label_str(
+        &mut self,
+        key: GenerationalId<StringId>,
+        val: GenerationalId<StringId>,
+    ) -> anyhow::Result<GenerationalId<LabelId>> {
+        let key = key.get(self.generation)?;
+        let val = val.get(self.generation)?;
+        let id = self.labels.dedup(Label::str(key, val));
+        Ok(GenerationalId::new(id, self.generation))
+    }
+
+    pub fn intern_labelset(
+        &mut self,
+        labels: &[GenerationalId<LabelId>],
+    ) -> anyhow::Result<GenerationalId<LabelSetId>> {
+        let labels = labels
+            .iter()
+            .map(|l| l.get(self.generation))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+        let labels = LabelSet::new(labels);
+        let id = self.label_sets.dedup(labels);
+        Ok(GenerationalId::new(id, self.generation))
+    }
```

Comment:
Using these functions skips the `Profile::validate_sample_labels` that the `add` would check before allowing them. I've noticed this because I had a test for this failure being handled correctly on the Ruby side, and the test did not see the error it expected.

Reply:
My feeling was that these checks should maybe be a debug_assert, but running them on every insert in prod is wasted CPU cycles